### PR TITLE
Get data for other profiles

### PR DESCRIPTION
**Pull request by aetheryna**
- Functionality for the API 🔌 
- Built routes for dynamic generation 🔨 
- Adjustments to the Navigation bar to lead to the logged-in user's profile 🔧

### CODE REVIEW

**Comment by bterone** [comment]
Could you rebase off the latest develop? 👍

File: pages/api/changeProfilePicture.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import multer from 'multer'
+
+const upload = multer({
+  storage: multer.diskStorage({
+    destination: './public/uploads/profile-pictures',
+    filename: (req, file, cb) => cb(null, file.originalname),
+  })
+})
+
+const fileUpload = upload.single('file')
```
Comment (bterone) [inline pages/api/changeProfilePicture.js]: This code is being duplicated from the `api/post.js`, is it possible to move this to some helper, that way we can pass into here some param and get the file upload directly like this 🤔 
```js
const fileUpload = saveToDisk('./public/uploads/profile-pictures');
```

**Comment by bterone** [comment]
Can you check if the activity feed is still working for this branch? It seems like its stopped working for me
Screen Shot 2021-08-23 at 8 42 10 AM

File: pages/api/getUserPosts.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import nc from 'next-connect'
+import prisma from '../../lib/adapters/prismaClient'
+
+const handler = nc()
+  .get(async(req, res) => {
+    const data = req.query
+
+    const users = await prisma.users.findMany({
```
Comment (bterone) [inline pages/api/getUserPosts.js]: I don't get this, we're searching for a specific ID, so do we need to use `findMany`?
Wouldn't `findUnique` suffice? 🤔

File: pages/profile/user/[uid].js [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import Head from 'next/head'
+import UserBanner from '../../../src/components/UserBanner'
+import UserProfileBar from '../../../src/components/UserProfileBar'
+import PostItem from '../../../src/components/PostItem'
+import UserProfileFriends from '../../../src/components/UserProfileFriends'
+import Feed from "../../../public/img/icons/feed.svg"
+import Friends from "../../../public/img/icons/friends.svg"
+import useUser from '../../../lib/auth/useUser'
+
+const userProfile = ({ userPosts }) => {
```
Comment (bterone) [inline pages/profile/user/[uid].js]: Shouldn't we be passing the user object instead? I'm a bit unclear of what is passing in here. 🤔

File: pages/api/getUserPosts.js (hunk 2 above)
Comment (aetheryna) [reply pages/api/getUserPosts.js]: Doesn't `findUnique` only return one post? 🤔 

REF : https://www.prisma.io/docs/concepts/components/prisma-client/crud#read

TLDR
image

File: pages/profile/user/[uid].js (hunk 3 above)
Comment (aetheryna) [reply pages/profile/user/[uid].js]: We are passing the `userPosts` prop which is being fetched from `getServerSideProps` on line 44, from there it does a check inside the page to see if the current user matches the user being fetched and renders the parts that's necessary for the current user such as Settings

File: pages/api/changeProfilePicture.js (hunk 1 above)
Comment (aetheryna) [reply pages/api/changeProfilePicture.js]: Done :smile:

File: pages/activity-feed/index.js [hunk 4]
```diff
@@ -17,7 +17,8 @@ const index = ({ postItems }) => {
 
 export const getStaticProps = async() => {
   const res = await fetch('http://localhost:3000/api/getAllPosts')
-  const postItems = await res.json()
+  const items = await res.json()
+  const postItems = items.findPosts
```
Comment (bterone) [inline pages/activity-feed/index.js]: Question here, why is `postItems` in an object like `{findPosts: {..}}`, what is in `items.findPosts`? 🤔

File: pages/api/getUserPosts.js (hunk 2 above)
Comment (bterone) [reply pages/api/getUserPosts.js]: Yes it does, but we're querying for a specific user here, right?
And there should only be one user we're returning right?

If we're searching for posts, is it possible to do as follows
```js
const posts = await prisma.posts.findMany({
   ...,
   where: {
     user_id: data.user_id
   }
```
So we're just returning the posts directly?
We can then include the user information we want with include 😄

Comment (bterone) [reply pages/api/getUserPosts.js]: From what I'm seeing, we're using this for the `userProfile`, so the API endpoint naming implies it's returning `Posts` primarily 😂 

I'm fine with calling this `getUser` so we know it's fetching the `User` primarily.
Also if that is the case, then we can use `findUnique`.
`include posts` is still part of the same user object so it should still be only one user 😄

File: pages/profile/user/[uid].js (hunk 3 above)
Comment (bterone) [reply pages/profile/user/[uid].js]: I've replied in this [comment](https://github.com/AetheryNA/crit/pull/18/files#r697514881),
I think after the changes in that, we can just pass in the `user` object directly 👍

File: pages/api/getUserPosts.js (hunk 2 above)
Comment (aetheryna) [reply pages/api/getUserPosts.js]: I'll change the name of the file so its `getUser` instead

**Comment by bterone** [comment]
> Can you check if the activity feed is still working for this branch? It seems like its stopped working for me
> Screen Shot 2021-08-23 at 8 42 10 AM

This seems to have been fixed 👍

File: pages/activity-feed/index.js (hunk 4 above)
Comment (aetheryna) [reply pages/activity-feed/index.js]: Made it more clearer

File: pages/activity-feed/index.js [hunk 5]
```diff
@@ -17,7 +17,8 @@ const index = ({ postItems }) => {
 
 export const getStaticProps = async() => {
   const res = await fetch('http://localhost:3000/api/getAllPosts')
-  const postItems = await res.json()
+  const getAllPosts = await res.json()
+  const postItems = getAllPosts.findPosts
```
Comment (bterone) [inline pages/activity-feed/index.js]: In the `getAllPosts.js`, line 19~21, currently we're calling like this
```js
return res.json({
  findPosts
})
```

What happens when we call it like below? Then I think we can get rid of this line here `const postItems = getAllPosts.findPosts` since we're just passing the returned value directly 🤔 
```js
return res.json(findPosts)
```

File: pages/api/getUser.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            user_id : true,
+            image_url : true,
+          }
+        }
+      },
+    })
+
+    return res.json({
+      users
+    })
```
Comment (bterone) [inline pages/api/getUser.js]: Similar comment here, can't we pass the user object directly? 
Then we can remove inside `[uid].js` line 12 `const userData = fetchedUserData.users`, since we're passing the object directly 🤔 

```suggestion
    return res.json(users)
```

**Review by bterone — COMMENTED** [review]
Just one more comment 👍

**Review by bterone — APPROVED** [review]
We have an issue with rendering a 404 if the user profile does not exist, I've created a bug ticket for that to work on later
